text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,356 @@
+    source: std::io::Error,
+    location: Location,
+    path: PathBuf,
+}
+
+impl fmt::Display for ExampleError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "Failed to read instrs from {}", path.display())
+    }
+}
+
+fn read_instrs(path: &Path) -> Result<String, ExampleError> {
+    std::fs::read_to_string(path).map_err(|source| {
+        ExampleError {
+            source,
+            path: path.to_owned(),
+            location: Location {
+                file: file!(),
+                line: line!(),
+            },
+        }
+    })
+}
+```
+
+Next we need to implement the `Error` trait to expose these members to the
+Error Reporter.
+
+```rust
+impl std::error::Error for ExampleError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.source)
+    }
+
+    fn provide_context(&self, type_id: TypeId) -> Option<&dyn Any> {
+        if id == TypeId::of::<Location>() {
+            Some(&self.location)
+        } else {
+            None
+        }
+    }
+}
+```
+
+And finally, we create an error reporter that prints the error and its source
+recursively along with the location data if it was gathered.
+
+```rust
+struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);
+
+impl fmt::Debug for ErrorReporter {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut current_error = Some(self.0.as_ref());
+        let mut ind = 0;
+
+        while let Some(error) = current_error {
+            writeln!(fmt, "    {}: {}", ind, error)?;
+
+            if let Some(location) = error.context::<Location>() {
+                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
+            }
+
+            ind += 1;
+            current_error = error.source();
+        }
+
+        Ok(())
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two additions necessary to the standard library to implement this
+proposal:
+
+Add a function for dyn Error trait objects that will be used by error
+reporters to access members given a generic type. This function circumvents
+restrictions on generics in trait functions by being implemented for trait
+objects only, rather than as a member of the trait itself.
+
+```rust
+impl dyn Error {
+    pub fn context<T: Any>(&self) -> Option<&T> {
+        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
+    }
+}
+```
+
+With the expected usage:
+
+```rust
+// With explicit parameter passing
+let spantrace = error.context::<SpanTrace>();
+
+// With a type inference
+fn get_spantrace(error: &(dyn Error + 'static)) -> Option<&SpanTrace> {
+    error.context()
+}
+```
+
+Add a member to the `Error` trait to provide the `&dyn Any` trait objects to
+the `context` fn for each member based on the type_id.
+
+```rust
+trait Error {
+    /// ...
+
+    fn provide_context(&self, id: TypeId) -> Option<&dyn Any> {
```

Comment:
It occurs to me reading this that it seems a little more common to name this kind of thing `get_context` in Rust but IDK.